semimpl/do_test.go

```diff
@@ -0,0 +1,177 @@
+			interval   = timeout / 5
+			numWorkers = 1
+			maxIter    = 0 // infinite iterations
+
+			margin      = 15 * time.Millisecond
+			maxDuration = timeout + margin
+		)
+
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+
+		gotDuration := timeFunc(func() {
+			semimpl.Do(ctx, numWorkers, maxIter, func() {
+				time.Sleep(interval)
+			})
+		})
+
+		if gotDuration > maxDuration {
+			t.Errorf(
+				"context timeout duration: exp < %dms, got %dms",
+				maxDuration.Milliseconds(), gotDuration.Milliseconds(),
+			)
+		}
+	})
+
+	t.Run("stop on context cancel", func(t *testing.T) {
+		var (
+			timeout    = 50 * time.Millisecond
+			interval   = 10 * time.Millisecond
+			numWorkers = 1
+			maxIter    = 0 // infinite iterations
+
+			margin      = 15 * time.Millisecond
+			maxDuration = timeout + margin
+		)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			time.Sleep(timeout)
+			cancel()
+		}()
+
+		gotDuration := timeFunc(func() {
+			semimpl.Do(ctx, numWorkers, maxIter, func() {
+				time.Sleep(interval)
+			})
+		})
+
+		if gotDuration > maxDuration {
+			t.Errorf(
+				"context cancel duration: exp < %dms, got %dms",
+				maxDuration.Milliseconds(), gotDuration.Milliseconds(),
+			)
+		}
+	})
+
+	t.Run("limit concurrent workers", func(t *testing.T) {
+		var (
+			numWorkers = 3
+			maxIter    = 12
+
+			durations                = make([]time.Duration, 0, maxIter)
+			maxIntervalWithinGroup   = 10 * time.Millisecond
+			minIntervalBetweenGroups = 30 * time.Millisecond
+		)
+
+		start := time.Now()
+		semimpl.Do(context.Background(), numWorkers, maxIter, func() {
+			durations = append(durations, time.Since(start))
+			time.Sleep(minIntervalBetweenGroups)
+		})
+
+		// check duration slice is coherent, grouping its values
+		// by expectedly similar durations, e.g.:
+		// 12 iterations / 3 workers -> 4 groups of 3 similar durations.
+		// With a callback duration of 30ms, we can expect something like:
+		// [[0ms, 0ms, 0ms], [30ms, 30ms, 30ms], [60ms, 60ms, 60ms], [90ms, 90ms, 90ms]]
+		// with a certain delta.
+		groups := groupby(durations, numWorkers)
+		for groupIndex, group := range groups {
+			// check durations within each group are similar
+			hi, lo := maxof(group), minof(group)
+			if interval := hi - lo; interval > maxIntervalWithinGroup {
+				t.Errorf(
+					"unexpected interval in group: exp < %dms, got %dms",
+					maxIntervalWithinGroup.Milliseconds(), interval.Milliseconds(),
+				)
+				t.Log(durations)
+			}
+
+			// check durations between distinct groups are spaced
+			if groupIndex == len(groups)-1 {
+				break
+			}
+			curr, next := minof(group), maxof(groups[groupIndex+1])
+			if interval := next - curr; interval < minIntervalBetweenGroups {
+				t.Errorf(
+					"unexpected interval between groups: exp > %dms, got %dms",
+					minIntervalBetweenGroups.Milliseconds(), interval.Milliseconds(),
+				)
+				t.Log(durations)
+			}
+		}
+	})
```

Comment:
[nitpick]
If I understand correctly, `durations` is the list of time at which each request has been done?
It is nitpicking, but isn't it a specific time rather than a duration (I get that this time = the duration from the start, but "durations" makes it sounds like it is the duration each request took, which is not the case here I think)? Could we rename it?

Reply:
I also went for `times` at first, but I was bothered having a `[]time.Duration` named `times`, which I expect to be `[]time.Time`.
We can still work on a more specific name, or at worst explicit it in a comment 🤷

Reply:
I don't mind having a `[]time.Duration` named `times` but I understand if you do 😁 
A comment also works if we do not come up with a more specific name.